docs/ADR-61-merkle-proofed-entities.md

```diff
@@ -0,0 +1,238 @@
+      "0x5f5170cdf5fdd7bb25c225d08b48361e41f05477880812f7f5954e75daa6c667",
+      "0x08ae25d236fa4105b2c5136938bc42f55d339f8e4d9feb776799681b8a8a48e7",
+      "0xadfcc425df780be50983856c7de4d405a3ec054b74020628a9d13fdbaff35df7",
+      "0xda4ee1c4148a25eefbef12a92cc6a754c6312c1ff15c059f46e049ca4e5ca43b",
+      "0x98c363c32c7b1d7914332efaa19ad2bee7e110d79d7690650dbe7ce8ba1002a2",
+      "0x0bd810301fbafeb4848f7b60a378c9017a452286836d19a108812682edf8a12a",
+      "0x1533c6b3879f90b92fc97ec9a1db86f201623481b1e0dc0eefa387584c5d93da",
+      "0x31c2c3dbf88646a964edd88edb864b536182619a02905eaac2a00b0c5a6ae207",
+      "0xc2088dbbecba4f7dd06c689b7c1a1e6a822d20d4665b2f9353715fc3a5f0d588",
+      "0x9e191109e34d166ac72033dce274a82c488721a274087ae97b62c9a51944e86f",
+      "0x5ff2905107fe4cce21c93504414d9548f311cd27efe5696c0e03acc059d2e445",
+      "0x6c764a5d8ded16bf0b04028b5754afbd216b111fa0c9b10f2126ac2e9002e2fa",
+    ],
+    hashingKeys: [
+      "id",
+      "name",
+      "description",
+      "image",
+      "thumbnail",
+      "data",
+      "i18n",
+      "createdAt",
+      "updatedAt",
+      "metrics",
+    ],
+    treeHashingMethod: "sha256",
+    entityHashingMethod: "IPFS",
+    entityHash: "bafkreifstqrxylmy6dwtw64ieyrgzv2sxrhh7awzb52bwn66utuo7327ae",
+  },
+}
+```
+
+### Proof generation
+
+The following is an example on how to generate a Merkle proof for a given entity.
+
+```typescript
+function ipfsHash(metadata, keys): string {
+  return hash(JSON.stringify(pick(metadata, keys)));
+}
+
+// Using the entity, the keys to be hashed and the other node hashes, build the merkle proof for the entity and return a new proofed entity.
+function buildEntityMetadataWithMerkleProof(baseEntity, keys, otherNodeHashes) {
+  const entityHash = ipfsHash(baseEntity, keys);
+  const { proof, index } = calculateProof(entityHash, otherNodeHashes);
+  return {
+    ...baseEntity,
+    merkleRoot: {
+      index,
+      proof,
+      hashingKeys: keys,
+      treeHashingMethod: "sha256",
+      entityHashingMethod: "IPFS",
+      entityHash,
+    },
+  };
+}
+```
+
+### Verification
+
+The following is an example on how to validate a Merkle proof for a given entity.
+
+```typescript
+function verifyHash(metadata, requiredKeys) {
+  // The keys used to create the hash MUST be present if they're required.
+  assert(
+    requiredKeys.every((key) => metadata.merkleRoot.hashingKeys.includes(key))
+  );
+  const generatedCrcHash = ipfsHash(metadata, metadata.merkleRoot.hashingKeys);
+  // The hash provided in the merkleProof for the entity MUST match the hash generated by the validator.
+  assert(metadata.merkleRoot.entityHash === generatedCrcHash);
+  // Based on the metadata.id, the merkle root is obtained from a trusted source i.e: the graph, or a smart contract directly
+  const merkleRoot = getMerkleRoot(metadata.id);
+  // Verify if the entity belongs to the Merkle Tree.
+  asset(
+    isProofValid(
```

Comment:
Just in case you want to add it. This should use the `verifyProof` from the [content hash tree lib](https://www.npmjs.com/package/@dcl/content-hash-tree)

Reply:
Changed it to use the same function found in the lib.